File: 4_Polymorphism/source/mallard.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "duck.h"
#include "mallard.h"
#include "mallard.r"

const char * colorNames[] = {"red", "brown", "white"};

static void mallardShow( Duck thisDuck );
static void _mallardMigrate( Mallard thisMallard );

static Mallard_Interface_Struct interface = {
    {.show=mallardShow},
    _mallardMigrate
};

Mallard
mallardCreate( void )
{
    Mallard newMallard = (Mallard)malloc(sizeof(Mallard_t));
    // TODO: Check for null pointer on malloc failure

    return newMallard;
}

void
mallardInit( Mallard thisMallard, char * name, featherColor color )
{
    printf("\tInitializing new mallard duck with name: %s\n", name);

    duckInit( &thisMallard->parentDuck, name );

    thisMallard->parentDuck.vtable = (Duck_Interface)&interface;
    thisMallard->myColor = color;
}

static void
mallardShow( Duck thisDuck )
{
    Mallard thisMallard = (Mallard)thisDuck;
    printf("\tHi! I'm a mallard duck. My name is %s. I have %s feathers.\n", thisMallard->parentDuck.name, colorNames[thisMallard->myColor]);
}

void
mallardMigrate( Mallard thisMallard )
{
    if ( thisMallard && thisMallard->parentDuck.vtable && ((Mallard_Interface)(thisMallard->parentDuck.vtable))->migrate )
    {
        ((Mallard_Interface)(thisMallard->parentDuck.vtable))->migrate(thisMallard);
    }
}

static void
_mallardMigrate( Mallard thisMallard )
{
    Duck thisDuck = (Duck)thisMallard;
    printf("\t%s: I'm migrating!\n", thisDuck->name);
}

File: Old/4_Run-time-Polymorphism_Inheritable/include/mallard.r
#ifndef MALLARD_R
#define MALLARD_R

#include "duck.r"
#include "mallard.h"

typedef struct Mallard_Interface_Struct const * Mallard_Interface;

typedef struct Mallard_t
{
    const Duck_t parentDuck;
    featherColor myColor;
} Mallard_t;

typedef struct Mallard_Interface_Struct
{
    Duck_Interface_Struct duckInterface;
    void (*migrate)( Mallard_t * thisMallard );
} Mallard_Interface_Struct;

void mallardInit( Mallard thisMallard, char * name, featherColor color );
void mallardDeinit( Duck thisDuck );

#endif //MALLARD_R